haystack/core/pipeline/pipeline.py

```diff
@@ -283,10 +283,10 @@ def run(self, word: str):
                         and last_waiting_for_input is not None
                         and before_last_waiting_for_input == last_waiting_for_input
                     ):
-                        # Are we actually stuck or there's a lazy variadic or a component with has only default inputs
-                        # waiting for input?
-                        # This is our last resort, if there's no lazy variadic or component with only default inputs
-                        # waiting for input we're stuck for real and we can't make any progress.
+                        if self._is_stuck_in_a_loop(waiting_for_input):
+                            # We're stuck! We can't make any progress.
+                            break
```

Comment:
This should be an exceptional event, no? Why aren't we throwing an error?

Reply:
We made the decision to always make `Pipeline.run()` return something when we implemented the current strategy. There could be that some parts of the `Pipeline` managed to run until the final output.

I'm not against making it fail but it's not something that I would do during the refactoring, I would consider it a breaking change.

Reply:
Let's emit a warning here, in that case. The user should be informed about it in some manner.